boot/boot.go

```diff
@@ -0,0 +1,171 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+)
+
+// A Device Boot Set is a collection of snaps that are involved in a
+// device's boot process.
+type DeviceBootSet interface {
+	// RemoveKernelAssets removes the unpacked kernel/initrd for the given
+	// kernel snap.
+	RemoveKernelAssets() error
+	// ExtractKernelAssets extracts kernel/initrd/dtb data from the given
+	// kernel snap, if required, to a versioned bootloader directory so
+	// that the bootloader can use it.
+	ExtractKernelAssets(snap.Container) error
+	// SetNextBoot will schedule the given OS or base or kernel snap to be
+	// used in the next boot. For base snaps it up to the caller to select
+	// the right bootable base (from the model assertion).
+	SetNextBoot() error
+	// ChangeRequiresReboot returns whether a reboot is required to switch
+	// to the given OS, base or kernel snap.
+	ChangeRequiresReboot() bool
+}
+
+// Lookup figures out what the right boot set is for the given
+// arguments, and returns it.
+func Lookup(s snap.PlaceInfo, t snap.Type, onClassic bool) (DeviceBootSet, error) {
+	if t != snap.TypeOS && t != snap.TypeKernel && t != snap.TypeBase && t != snap.TypeSnapd {
+		return nil, fmt.Errorf("cannot lookup boot set with snap %q of type %q", s.SnapName(), t)
```

Comment:
I would have thought this would return nil or a detectable error, or nop implementation if the type is not meaningful

Reply:
hm, a nop implementation might be good. We have tests that want the error I think but I'll look into it.